File: binarysearches.cpp
#include <bits/stdc++.h>

using namespace std;

int first_greq(vector<int>& x, int target){
    int left=0,right=x.size()-1;
    while(left<right){
        int mid=left+(right-left)/2;
        if(x[mid]>=target){
            /*
             * Since our middle element
             * is greater than or equal to target,
             * it is a possible candidate, but
             * that does not rule out the elements
             * to the left of the middle from being
             * candidates as well.
             */
            right=mid;
        }else{
            /*
             * At this point, we know x[mid]<target,
             * which disqualifies it from
             * being a candidate; we're looking for
             * the first element >= to target, so
             * we must move our left bound up one.
             */
            left=mid+1;
        }
    }
    return x[left]>=target?left:-1;
}
int first_gr(vector<int>& x, int target){
    int left=0,right=x.size()-1;
    while(left<right){
        int mid=left+(right-left)/2;
        if(x[mid]>target){
            // logic from first_greq
            // applies here, just change
            // >= operator to >
            right=mid;
        }else{
            // consult previous comment
            left=mid+1;
        }
    }
    return x[left]>target?left:-1;
}

File: binarysearches.py
# lower bound implementation
# for an array sorted by the
# comparator function `cmp`
def first_greq(arr,target,cmp):
    i,j = 0,len(arr)-1
    while i < j:
        m = i + (j-i) // 2
        if cmp(arr[m], target) >= 0:
            j = m
        else:
            i = m + 1
    if cmp(arr[i], target) < 0: return i + 1
    return i

# upper bound implementation
# for an array sorted by the
# comparator function `cmp`
def first_gr(arr,target,cmp):
    i,j = 0,len(arr)-1
    while i < j:
        m = i + (j-i) // 2
        if cmp(arr[m], target) > 0:
            j = m
        else:
            i = m + 1
    if cmp(arr[i], target) <= 0: return i + 1
    return i
